src/mon/Monitor.h

```diff
@@ -345,7 +343,21 @@ class Monitor : public Dispatcher,
     ScrubState() : finished(false) { }
     virtual ~ScrubState() { }
   };
-  std::shared_ptr<ScrubState> scrub_state; ///< keeps track of current scrub
+
+  struct ScrubContext {
+    ScrubState scrub_state;       ///< keeps track of current scrub
+    version_t scrub_version;      ///< paxos version we are scrubbing
+    std::map<int,ScrubResult> scrub_result;  ///< result so far
+    ScrubContext() {
+      scrub_version = 0;
+      scrub_result.clear();
+    }
+    ~ScrubContext() {
+      scrub_version = 0;
+      scrub_result.clear();
+     }
+  };
+  std::atomic<std::shared_ptr<ScrubContext>> scrub_con; ///< keeps track of scrub_context
```

Comment:
Hey @mohit84 there is a build failure related to this line:

```
/usr/include/c++/11/atomic: In instantiation of ‘struct std::atomic<std::shared_ptr<Monitor::ScrubContext> >’:
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.3.0-5886-gc279fd8b/rpm/el9/BUILD/ceph-19.3.0-5886-gc279fd8b/src/mon/Monitor.h:360:46:   required from here
/usr/include/c++/11/atomic:211:21: error: static assertion failed: std::atomic requires a trivially copyable type
  211 |       static_assert(__is_trivially_copyable(_Tp),
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11/atomic:211:21: note: ‘false’ evaluates to false
make[2]: *** [src/mon/CMakeFiles/mon.dir/build.make:398: src/mon/CMakeFiles/mon.dir/NVMeofGwMap.cc.o] Error 1
```

Reply:
Build ref: https://shaman.ceph.com/builds/ceph/wip-yuri3-testing-2024-10-31-0832/
Testing ref: https://tracker.ceph.com/issues/68795

Reply:
As per error it seems std::atomic only works with trivially copyable type and std::shared_ptr is not trivially copyable. As per cpp reference for c++20 it should work fine but not sure  why it is not working?

@rzarzynski 
Do you think we need to use mutex instead of using atomic to avoid the same?

Reply:
I think the reason is buggy / laggy compiler.

```
-- The CXX compiler identification is GNU 11.2.0
-- The C compiler identification is GNU 11.2.0
...
/usr/include/c++/11/atomic: In instantiation of 'struct std::atomic<std::shared_ptr<Monitor::ScrubContext> >':
/build/ceph-19.3.0-5886-gc4040e92/src/mon/Monitor.h:360:46:   required from here
/usr/include/c++/11/atomic:211:21: error: static assertion failed: std::atomic requires a trivially copyable type
  211 |       static_assert(__is_trivially_copyable(_Tp),
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

godbolt.org shows GCC 11 is unable to compile even:

```cpp
std::atomic<std::shared_ptr<int>> atomic_ptr_test;
```

complaining about non-trivial copy-constructibility. I bet the reason is GCC simply lacks the partial specialization for `shared_ptr<T>`. GCC 12 is fine.

The workaround could be https://www.boost.org/doc/libs/1_75_0/libs/smart_ptr/doc/html/smart_ptr.html#atomic_shared_ptr

Reply:
Thanks, I have changed the code and use boost::atomic_shared_ptr.